### PR TITLE
Implement has_database_privilege function

### PR DESCRIPTION
**Pull request by matriv**
## Summary of the changes / Why this improves CrateDB

See commit messages.
Closes: https://github.com/crate/crate/issues/10251

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by matriv** [comment]
For Postgres, user must be granted `create on database` to return `true` for `has_database_privilege(<user>, <db>, 'CREATE')`:
```
lala=> select has_database_privilege('lala', 'create');
 has_database_privilege 
------------------------
 f
(1 row)

GRANT CREATE ON database lalaDB TO lala; -- as psql user

lala=> select has_database_privilege('lala', 'create');
 has_database_privilege 
------------------------
 t
(1 row)
```
For CrateDB I choose to check for `DDL` on any schema, but maybe we want to check `DDL` on `CLUSTER` to show that the user has the ability to create schema, and not objects (functions, relations, views) within it. WDYT?

**Comment by matriv** [comment]
@BaurzhanSakhariev @seut @proddata 

For Postgres, user must be granted `create on database` to return `true` for `has_database_privilege(<user>, <db>, 'CREATE')`:
```
lala=> select has_database_privilege('lala', 'create');
 has_database_privilege 
------------------------
 f
(1 row)

GRANT CREATE ON database lalaDB TO lala; -- as psql user

lala=> select has_database_privilege('lala', 'create');
 has_database_privilege 
------------------------
 t
(1 row)
```
For CrateDB I choose to check for `DDL` on any schema, but maybe we want to check `DDL` on `CLUSTER` to show that the user has the ability to create schema, and not objects (functions, relations, views) within it. WDYT?

**Comment by proddata** [comment]
> For CrateDB I choose to check for `DDL` on any schema, along `DDL` on `CLUSTER`, but maybe we want to only check for the latter, to reflect that the user has the ability to create schemas, and not objects (functions, relations, views) within a schema. WDYT?

I would tent to be less "restrictive". In CrateDB it is possible to grant `DDL` permissions for not yet existing schemas/objects.

```sql
CREATE USER lala;
GRANT DDL ON SCHEMA new_lala TO lala;
```

which allows the user `lala` to implicitly create the schema `new_lala` by creating any database object within that schema.

```
cr> create user lala;                                                                                                          
CREATE OK, 1 row affected  (0.091 sec)

cr> grant ddl on schema new_lala to lala;                                                                                      
GRANT OK, 1 row affected  (0.069 sec)

cr> SELECT nspname FROM pg_catalog.pg_namespace; 
                                                                                                                               
+--------------------+
| nspname            |
+--------------------+
| blob               |
| information_schema |
| devices            |
| pg_catalog         |
| doc                |
| sys                |
+--------------------+
SELECT 6 rows in set (0.013 sec)

cr> CREATE TABLE new_lala.lala_table (txt TEXT);                                                                               
CREATE OK, 1 row affected  (0.488 sec)

cr> SELECT nspname FROM pg_catalog.pg_namespace; 
                                                                                                                               
+--------------------+
| nspname            |
+--------------------+
| blob               |
| information_schema |
| devices            |
| pg_catalog         |
| doc                |
| sys                |
| new_lala           |
+--------------------+
SELECT 7 rows in set (0.005 sec)
```

**Comment by matriv** [comment]
> > For CrateDB I choose to check for `DDL` on any schema, along `DDL` on `CLUSTER`, but maybe we want to only check for the latter, to reflect that the user has the ability to create schemas, and not objects (functions, relations, views) within a schema. WDYT?
> 
> I would tent to be less "restrictive". In CrateDB it is possible to grant `DDL` permissions for not yet existing schemas/objects.

So, as it's in the current status, `DDL` on `CLUSTER`, or `DDL` on any schema, right?

File: server/src/main/java/io/crate/expression/scalar/HasDatabasePrivilegeFunction.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    @Nullable
+    protected Collection<Privilege.Type> parsePrivileges(String privilege) {
+        Collection<Privilege.Type> toCheck = new HashSet<>();
+        String[] privileges = privilege.toLowerCase(Locale.ENGLISH).split(",");
+        for (String p : privileges) {
+            p = p.trim();
+            switch (p) {
+                case "connect" -> toCheck.add(Privilege.Type.DQL);
```
Comment (proddata) [inline server/src/main/java/io/crate/expression/scalar/HasDatabasePrivilegeFunction.java:97]: Don't know how this is handled internally, but in theory a user that has no explicit privileges granted, still has access to some `information_schema` and `pg_catalog` tables and therefore could connect (and query) the database

Comment (matriv) [reply server/src/main/java/io/crate/expression/scalar/HasDatabasePrivilegeFunction.java:97]: This is a mapping we do ourselves, there is no connect privilege for CrateDB, so mapping it to DQL, Also mapping the rest `TEMP`, and `CREATE` to the available `DML` and `DDL` enum values we have.
When we validate here: https://github.com/crate/crate/pull/13377/files#diff-2328b4da2125364630421b4cb3da5089ba02a5638fd0e83b1520740b07786099R65 we check for DQL => connect, if the user exists it's true, for `DML` => temp always false, we don't have temp tables in CrateDB, for `DDL`=> create, we check AL on cluster of DDL on any schema.

File: docs/general/builtins/scalar-functions.rst [hunk 2]
```diff
@@ -3401,6 +3401,52 @@ Example::
...
+
+:CONNECT:
+  is ``true`` for all defined users in the database
+
+:CREATE:
+  is ``true`` if the user has any ``DDL`` privilege on ``CLUSTER`` or on any
+  ``SCHEMA``
+
+:TEMP:
+  if ``false`` for all users
```
Comment (BaurzhanSakhariev) [inline docs/general/builtins/scalar-functions.rst:3436]: typo? should be `is`?

**Review by BaurzhanSakhariev — APPROVED** [review]
Left some minor comments, LGTM.


Thanks for refactoring!

File: docs/general/builtins/scalar-functions.rst (hunk 2 above)
Comment (matriv) [reply docs/general/builtins/scalar-functions.rst]: yes, thank you!

**Pull request merged**